src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,998 @@
+#include "Domain/ElementId.hpp"  // IWYU pragma: keep
+#include "Domain/Mesh.hpp"
+#include "Domain/OrientationMap.hpp"  // IWYU pragma: keep
+#include "Domain/Tags.hpp"            // IWYU pragma: keep
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Math.hpp"
+#include "Utilities/TMPL.hpp"
+
+// IWYU pragma: no_include <algorithm>
+// IWYU pragma: no_forward_declare Variables
+
+/// \cond
+namespace SlopeLimiters {
+template <size_t VolumeDim, typename TagsToLimit>
+class Hierarchical;
+}  // namespace SlopeLimiters
+/// \endcond
+
+namespace SlopeLimiters {
+enum class HierarchicalType { Krivodonova };
+
+std::ostream& operator<<(std::ostream& os, HierarchicalType t) noexcept;
+
+/*!
+ * \ingroup SlopeLimitersGroup
+ * \brief A general hierarchical slope limiter (HL) implementation.
+ *
+ * Implements HL's that limit the logical derivatives of the `Tags` starting at
+ * the highest derivatives. The simplest of these limiters is described in \cite
+ * Krivodonova2007. The Krivodonova limiter is applied to the spectral
+ * coefficients. In 3d the function being limited is expanded as:
+ *
+ * \f{align}{
+ * u^{l,m,n}=\sum_{i,j,k=0,0,0}^{N_i,N_j,N_k}c^{l,m,n}_{i,j,k}
+ *  P_{i}(\xi)P_{j}(\eta)P_{k}(\zeta)
+ * \f}
+ *
+ * where \f$\left\{\xi, \eta, \zeta\right\}\f$ are the logical coordinates,
+ * \f$P_{i}\f$ are the Legendre polynomials, and the superscript \f$\{l,m,n\}\f$
+ * represents the element indexed by \f$l,m,n\f$. The coefficients are limited
+ * according to:
+ *
+ * \f{align}{
+ * \tilde{c}^{l,m,n}_{i,j,k}=\mathrm{minmod}
+ *   &\left(c_{i,j,k}^{l,m,n},
+ *          \alpha_i\left(c^{l+1,m,n}_{i-1,j,k}-c^{l,m,n}_{i-1,j,k}\right),
+ *          \alpha_i\left(c^{l,m,n}_{i-1,j,k}-c^{l-1,m,n}_{i-1,j,k}\right),
+ *     \right.\notag \\
+ *   &\;\;\;\;
+ *          \alpha_j\left(c^{l,m+1,n}_{i,j-1,k}-c^{l,m,n}_{i,j-1,k}\right),
+ *          \alpha_j\left(c^{l,m,n}_{i,j-1,k}-c^{l,m-1,n}_{i,j-1,k}\right),
+ *     \notag \\
+ *   &\;\;\;\;\left.
+ *          \alpha_k\left(c^{l,m,n+1}_{i,j,k-1}-c^{l,m,n}_{i,j,k-1}\right),
+ *          \alpha_k\left(c^{l,m,n}_{i,j,k-1}-c^{l,m,n-1}_{i,j,k-1}\right)
+ *     \right),
+ * \label{eq:krivodonova 3d minmod}
+ * \f}
+ *
+ * where \f$\mathrm{minmod}\f$ is the minmod function defined as
+ *
+ * \f{align}{
+ *  \mathrm{minmod}(a,b,c)=
+ *  \left\{
+ *  \begin{array}{ll}
+ *    \mathrm{sgn}(a)\min(\lvert a\rvert, \lvert b\rvert,
+ *    \lvert c\rvert) & \mathrm{if} \;
+ *    \mathrm{sgn}(a)=\mathrm{sgn}(b)=\mathrm{sgn}(c) \\ 0 & \mathrm{otherwise}
+ *  \end{array}\right.
+ * \f}
+ *
+ * Krivodonova \cite Krivodonova2007 requires \f$\alpha_i\f$ to be in the range
+ *
+ * \f{align*}{
+ * \frac{1}{2(2i-1)}\le \alpha_i \le 1
+ * \f}
```

Comment:
She doesn't do anything in 3D, she only does 2D. Maybe I'm  confused about this, but our Legendre polynomials are always normalized to P(1)=1 independent of dimension, so would that change the case in a tensor product basis?

Reply:
Krivodonova normalizes her 2D (you're right, there is no 3D) Legendre polynomials so that `integral P_i P_j = delta_ij`. This is a different normalization, and leads to a different set of bounds on alpha. As a result, your documentation does not obviously match the paper when dim > 1, and you need to document this difference, so the reader can usefully compare your code to the paper. Add a note saying, roughly:

"Note that we normalize our Legendre polynomials by P(1) = 1; this is the normalization Krivodonova uses in 1D, (but not in 2D), which is why our bounds on alpha match Krivodonova Eq. 14 (but not Eq. 23)."

Reply:
Okay, done.